rosidl_cmake/cmake/rosidl_generate_interfaces.cmake

```
@@ -137,22 +137,50 @@ macro(rosidl_generate_interfaces target)
   # afterwards all remaining interface files are .idl files
   list(APPEND _idl_tuples ${_idl_adapter_tuples})
 
-  # to generate action interfaces, we need to depend on "action_msgs"
+  set(_service_msgs_added FALSE)
+  set(_action_msgs_added FALSE)
   foreach(_tuple ${_interface_tuples})
     string(REGEX REPLACE ".*:([^:]*)$" "\\1" _tuple_file "${_tuple}")
     get_filename_component(_parent_dir "${_tuple_file}" DIRECTORY)
-    if("${_parent_dir}" STREQUAL "action")
+
+    if("${_parent_dir}" STREQUAL "action" AND NOT _action_msgs_added)
```

```suggestion
    if(NOT _action_msgs_added AND "${_parent_dir}" STREQUAL "action")
```

may be more "efficient", as the STREQUAL part doesn't need to be evaluated if the flag is true

---

See [91ea5a9](https://github.com/ros2/rosidl/pull/700/commits/91ea5a9062cecf44e8d77fc730c06ea89e922593)